README
musync README

    authors John-John Tedro (johnjohn.tedro@gmail.com) a.k.a udoprog
    contact: mail or #ost@tinfoilnet.org

    SEE THE ChangeLog for important changes that might effect you!

    For windows users: you are on your own atm, there simply is no way to get an 'ok' working ui under windows.
                       in future we might add a few .bat scripts but plans are being laid for gui-wrapper.
                       I have added binaries for cp, md5sum and sed. These are the result of
                       http://unxutils.sourceforge.net/ so if you wan't latest version and such you should look there.

    === Dependencies ===
        * mutagen >=1.12
          http://www.sacredchao.net/quodlibet/wiki/Download
        * the common *nix tools cp, rm and sed (GNU sed is recommended)

    === Installation ===
        posix: run 'python setup.py install' as superuser (or build as normal, then install)
        
        read what it says.

        example configurations will be in:
        posix: /usr/local/musync
        windows: c:\temp

    === Configuration ===
        Musync's default configuration will not work for your system, you have to manually configure it.


    === General ===
        to use musync for adding music to your filestructure, do something like:
            
            musync -R add /path/to/album/

        or for piping:
            
            ls /path/to/music/*/* | musync add

        piping should be used in second hands since it doesn't allow musync an overview of the directories.

        deleting has similar syntax, just use 'rm'
        
        If musync complains about bad metadata, check out the log-file that was created (default /tmp/log-fixes.log)
        and run it trough some nice tagger, like musicbrainz picard.

        there is a manual in construction at svn which can be found on http://sf.net/projects/musync 

        --- what is 'fix'? ---
            i noticed that my music files really got screwed around when adding stuff to musync.sed for naming rules, and i had
            to remove a directory which hade been created by a bad rule etc.

            what 'fix' does is that it sanitychecks the files in your repository, and makes sure that no empty directories and
            files tagged differently exists.
            directories will be removed, files will just be moved.

            There is quite extensive output using this command.

    === Configuration ===
        MuSync will look for the following configuration files:
            
            /etc/musync.conf
            ~/.musync
        
        It uses the native Python RawConfigParser, which accepts the following syntax:

            [general]
            root: /path/to/foo
            root= /path/to/foo
            # I am a comment
            
        Everythin written in a section [foo] will be used by MuSync when using '-c foo', this will override options.
        The [general] section is the default section, this will always be read.

        If more than one section are read, the keys are overwritten cronologically, starting with the first read.
        this will result in the 'last' key read will be applied to the program.

    === Notes ===
        by default, the files are placed in

            %root%/%artist%/%album%/%track%-%title%.%ext%

        sed-scripting knowledge if a must if you dislike the way musync sanitizes string names.
        But if you find the interest, check the 'sed-with' option in configuration
        (default /etc/musunc.conf) and tamper with sed-naming file (default /etc/musync.sed).
        
        We stress that musync has _only_ been developed and tested under utf-8 locale system, but should be able to run properly on others as well.

        MuSync could be made operating system independant but would require some skills configuring sed and using configuration, we'll se in the future.
    
    === IMPORTANT ===
        Current behaviour when character with decimal value > 127 is encountered, it will be replaced by a series of 0xXX - hexadecimal values.
        example (if letter would not have been supported):
            Random Artist - Hallå Världen! -> random_artist - hall0xXX0xXX_v0xXX0xXXrlden
            
        This is to make sure that on cases where naming fails, it will be noticed.
        A warning will also be printed.

        a way to search for misnaming is to run the following command in you music library:

            ls -R | grep -r "0x[0-9a-e][0-9a-e]"

        Look out for these in your libraries and add them to /etc/musync.sed for proper replacing rules. Remember to comment and try to be as international as possible.
        - If you find files which are named incorrectly, please do try to fix the script musync.sed (default /etc/musync.sed) to name correctly.
        - If you feel that you have a fix that is important please send it to us!


ChangeLog
Fri Nov 28 20:26:33 CET 2008 by udoprog
	* fixed minor issues with lockfiles dissapearing.
	* added several configuration warnings.

Thu Oct  9 18:57:08 CEST 2008 by udoprog
	* added support for mp4 and ape in metadata reading,
	did a rework to make it more flexible.

Thu Oct  9 05:36:02 CEST 2008 by plastbunker
	* added year information
	* fixed broken termcaps

Thu May  8 03:49:53 CEST 2008 by udoprog
	* coloring no longer does anything, since terminal capabilites will check
	if coloring is possible. if not it is just ignored. there was also some
	dofficulties involved with this that wasn't expected, and the simple
	sollution was this.
	* Progressbar now works, all output is written to log.


Tue May  6 12:33:59 CEST 2008 by udoprog
	* added experimental transcoding options. They are explained properly when
	running 'musync --help' and trying the '-T' option.
	-T flac=ogg would try to (encode) flac-filer into ogg. Musync would look
	for the key 'flac-to-ogg' in configuration and expects the value of the
	key to be a command using the variables %(source)s %(dest)s for encoding
	the flac file (as source) and the target ogg (as dest).
	* fixed checking for similar files in music database, files named the same
	but with different extensions are effected. Users will be prompted to use
	--allow-similar to allow for example an ogg and a flac file with same
	artist, title, album and track to be added to db.
	* all extensions in configuration no longer includes a dot '.', it is
	important to fixed the keys 'supported-ext' and 'format' in configuration.
	This has been fixed in example configuration.
	* added terminal capabilities, if curses exists on target computer.
	Coloring is now controlled trough this. (more dynamic).

Sun May  4 17:39:54 CEST 2008 by udoprog
	* removed key sed-naming and renamed sed-with with filter-with. There is
	also the possibility to use the %(field)s variable in the string which
	enables you to use different configs for different fields in the metadata
	(artist, album ...).

Fri May  2 22:38:39 CEST 2008 by udoprog
	* added bash_completion script for easier usage. fix/lock/unlock command
	lists database instead of working directoriy.
	* fixed minor bugs, check all operations. now prints relativa paths (to
	root) instead of long absolute ones.

Wed Apr 30 00:09:06 CEST 2008 by udoprog
	* added better help when configuration is faulty (thanks to talyz for
	this)
	* ive also started applying a better ChangeLog syntax.
	* added option --progress for a progress-meter instead of output.
	* santitized some spillage output. (--help, --version, etc...)

2008.04.23 udoprog
	* added hinting

earlier
	* Added option --overwrite (now known as --force)
	* Fixed bug with absolute path reference in MusicFile. It used to disrupt
	symlinking.
	* Fixed comma-seperated value acceptance for --config parameter.
	* Fixed internals and added class Withs (for fun) to handle commands.
	* Change decoding to utf-16 instead of utf-8 when cleaning names. (major)
	* Warnings are now printed onto stderr.
	* Fixed recognition of 'tracknumber' key as 'track'
	* Added key 'coloring' and 'default-config' (-C/--coloring in cli).
	* Warnings are no longer logged.
	* Stability and naming conventions somewhat differs.
	* Linear reference is now possible with configurations.
	* Musync has been divided into several modules, see for your self. Global
	reconstructure and reuse of code, now it is neat and tidy.
	* you should no longer use 'mv' with add-with, since it breaks
	hash-testing.
	* Fixing now works, see README for what it does.
	* supported-ext has new syntaX, REQUIRES dots as prefix (like
	'.mp3,.flac')
	* overwrite is now known as force. -O is -F. (also true in configuration).
	* Fixed windows compatibility. Problems with how to code default config
	file remains.
	* Fixed install scripts.
	* added locking/unlocking capabilities
	* fixed -M properly
	* added --lock (for locking when adding/fixing)
	* fixed usage of gnu_getopt instead of getopt.
	* added hinting
	* Changed output when file is added or removed to include basename.
	* Configuration reading is performed by an automatic loop.
	* Configuration is read in "/.musync
	* Peronalized configurations are read, named by [section], used with '-c'


musync/subp.py
#
# Musync subprocess - A few nifty subprocessing hacks.
#
# Yes, this enables the program to simplify interprocessing.
#
# author: John-John Tedro <pentropia@gmail.com>
# version: 2008.1
# Copyright (C) 2007 Albin Stjerna, John-John Tedro
#
#    This file is part of Musync.
#
#    Musync is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    Musync is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with Musync.  If not, see <http://www.gnu.org/licenses/>.
#

from musync.opts import Settings;
import subprocess as sp;
from musync.errors import WarningException, FatalException;
import musync.printer as Printer;

def filter_with(str, field):
    """
    Wrapper for use with the key 'filter-with'.
    """
    
    filter_cmd = Settings["filter-with"].split(' ');
    filter_cmd = [ x%{'field': field} for x in filter_cmd ];
    return safe_child(sc_filter_with, [filter_cmd, str]);

def sc_filter_with(args):
    """
    'filter-with' to be used as first argument to safe_child.
    """
    proc = sp.Popen( args[0], stdin=sp.PIPE, stdout=sp.PIPE );
    proc.stdin.write( args[1] );
    proc.stdin.close();
    filter_data = proc.stdout.read();
    state = proc.wait();
    return (state, filter_data);

def transcode_with(cmd, source, dest):
    trans_cmd = cmd.split(' ');
    trans_cmd = [x%{ 'source': source, 'dest': dest} for x in trans_cmd];
    return safe_child(sc_trans_with, trans_cmd);

def sc_trans_with(args):
    proc = sp.Popen(args);
    state = proc.wait();
    return (state, state);

def add_with( source_file, destination_file ):
    add_cmd = Settings["add-with"].split(' ');
    add_cmd = [ x%{ 'source': source_file, 'dest': destination_file } for x in add_cmd ];
    return safe_child(sc_add_with,add_cmd);

def sc_add_with(args):
    proc = sp.Popen( args );
    state = proc.wait();
    return (state, state);

def rm_with(target):
    rm_cmd = Settings["rm-with"].split(' ');
    rm_cmd = [ x%{ 'target': target } for x in rm_cmd ];
    return safe_child(sc_rm_with, rm_cmd);

def sc_rm_with(args):
    proc = sp.Popen( args );
    state = proc.wait();
    return (state, state);

def hash_with(target):
    """
    Uses Hash with.
    """
    hash_cmd = Settings["hash-with"].split(' ');
    hash_cmd = [ x%{ 'target': target } for x in hash_cmd ];
    return safe_child(sc_hash_with, hash_cmd);

def sc_hash_with(args):
    """
    'hash-with' wrapper to be used as first argument to safe_child.
    """
    proc = sp.Popen(args, stdin=None, stdout=sp.PIPE);
    hash_data = proc.stdout.read();
    proc.stdout.close();
    state = proc.wait();
    return (state, hash_data.split(' ')[0]);

def safe_child(call, args):
    """
    Ok, this is officially the weirdest thing ive ever done in python.
    Now, simple enough, if anything goes wrong with the call, we wan't to try again,
    claiming that the process has been 'shaked'. This is to make sure the main process
    signals doesn't flow over to the subprocess as is default behavior.
    
    Practically this means the subprocess ignores an interrupt by retrying.

    This really is a problem when python relays interrupts at a bad timing to the child process.
    """
    #Utilise the supprocessing module
    state = 1;
    i = 0;
    ret = None;
    while True:
        if i > 0:
            raise FatalException("child process failed to many times");
        
        try:
            (state, ret) = call(args);
        except Exception, e:
            print e;
            state = 1;
        
        if state != 0:
            Printer.warning( "shaked process %s, %s"%(repr(call), repr(args)) );
            i += 1;
            continue;

        break;
    return ret;

def utf162ascii(text):
    """
    Create pure ascii strings from utf-16 strings with escaped to long chars.
    """
    safe_text = "";
    i = 0;
    while i < len( text ):
        first = ord( text[i] );
        second = ord( text[i+1] );

        if first is 0 and second <= 127:
            safe_text += chr( second );
        else:
            safe_text += str( hex( first ) ) + str( hex( second ) );

        i += 2;

    return safe_text;

def sanitize_with_filter(text, field):
    """
        remove none-filename-ish characters from strings.
        Check configuration to change this behaviour.
    """
    
    all = "";
    i = 0;
    back = None;
    for c in text.encode( "utf-16" ):
        if i < 2:
            i += 1;
            continue;
        if i % 2 is 0:
            back = ord(c);
        else:
            all += chr( ord(c) );
            all += chr( back );
        i += 1;
    
    filter_text = filter_with(all, field);
    
    safe_text = "";
    next = False;
    prev = 0;
    
    # Since we are working with 2-byte per char strings, it must always be possible to divide the strings without rest with 2
    if len( filter_text ) % 2 is not 0:
        raise FatalException( "musync.filter corrupted string" );
    
    safe_text = utf162ascii( filter_text );
    return safe_text;
